Rake task for managing plugins

jonasheinrich opened the pull request:
What we plan on doing:
- put steps into transactions
- include more sanity checks (e.g. a plugin is only listed once)
- introduce a blacklist-key for every plugin

What we might do:
- split the task up for developers
- rake plugin:add[community]

Comments welcome!

This is the relevant work package: https://community.openproject.org/work_packages/18720

Commit:
First draft of rake task for managing plugins

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://jonasheinrich-feature-rake-tasks-for-administering-plugins.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://jonasheinrich-feature-rake-tasks-for-administering-plugins.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Commit:
Refactors some methods into the class Plugin

Commit:
Some refactorings

jonasheinrich commented:
@opf/developers Comments welcome!

Commit:
Adds dependency check on remove, checks for plugins.yml and fixes a bug

Commit:
Some style flaws hound noticed

Commit:
Removes unnecessary comments

[ci skip]

NobodysNightmare commented on lib/tasks/plugins.rake:
```diff
@@ -88,3 +88,17 @@ namespace :redmine do
 end
+
+namespace :plugins do
+desc 'Adds Meeting plugin for now'
```
Are the descriptions still valid?

I'll have to do further code lookup, but I hope that this will do more than add/remove the _Meeting_ plugin.

jonasheinrich commented:
This is the relevant work package: https://community.openproject.org/work_packages/18720

linki commented on lib/plugin_manager.rb:
```diff
@@ -0,0 +1,209 @@
+  end
+
+  def _migrate
+    system "RAILS_ENV='#{@environment}' rake db:migrate"
```
how about using `bundle exec` to run rake?

linki commented on lib/plugin_manager.rb:
```diff
@@ -0,0 +1,209 @@
+  end
+
+  def self._available_plugins
+    @available_plugins || _load_available_plugins
```
most people use `@available_plugins ||= _load_available_plugins`. you can drop the instance var assignment in `_load_available_plugins` then.

linki commented on lib/plugin_manager.rb:
```diff
@@ -0,0 +1,209 @@
+
+  def initialize(name)
+    if name == 'pdf-inspector'
+      @name = 'pdf-inspector'
```
could you please add a comment on why this is necessary and when it can be removed.

linki commented on lib/plugin_manager.rb:
```diff
@@ -0,0 +1,209 @@
+        exit
+      end
+      @name = name
+      @url = Plugin._available_plugins[name][:url]
```
either you're calling a private method here or it shouldn't begin with an underscore in the first place.. pick one :)

linki commented on lib/plugin_manager.rb:
```diff
@@ -0,0 +1,205 @@
+    end
+  end
+
+  def _not_needed_by_any_other_than?(plugin)
```
:+1: @houndci

linki replied:
or maybe not... `plugin` is used in the method ^^

linki commented:
please add tests :racehorse:

Commit:
Use bundle exec to rake

Commit:
Drop the unnecessary instance variable assignment

Commit:
Clarifies comments about pdf-inspector

Commit:
Makes method private

[ci skip]

Commit:
Fixes an error with the place of a constant

Commit:
Updates description of the rake tasks

Commit:
Adds sorting

Commit:
Adds a todo for duplicated lines

Commit:
Updates available plugins

[ci skip]

Commit:
Updates comments

Commit:
Removes duplicated lines

Commit:
Adds check for duplicated plugins

[ci skip]

Commit:
Moves the anomalies of pdf-inspector to plugins.yml

Commit:
Makes the caching of available plugins work

Commit:
Makes a method of Plugin public

Commit:
Extends error message

Commit:
Extends further error messages

[ci skip]

Commit:
Adds tests

Commit:
Fixes deletion of dependencies

Commit:
Adds and slightly refactors tests

Commit:
Refactors generation of Plugins with list of names

Commit:
Just style fixes

[ci skip]

Commit:
Fixes generation of gemfile with versions

Commit:
Fixes dependencies of pdf_export

[ci skip]

Commit:
Fixes new lines

Commit:
Moves costs before reporting

Commit:
Adds tag and commit as possible keys

[ci skip]

Commit:
Updates plugins' branch to dev

Commit:
Don't delete plugins that could have been installed before

[ci skip]

Commit:
Makes publicly used functions public

[ci skip]

ulferts commented:
@jonasheinrich this is labeled 'WIP' is that correct?

kgalli commented:
@ulferts Yes it still is. I suggested to change the task to use plugins only available on rubygems for now. I will update the plugins/gems on rubygems.org and adjust the rake task accordingly. Otherwise we will have a lot of trouble regarding dependencies of plugins.

linki commented:
**tl/dr** check in `Gemfile.lock`, `schema.rb`, etc. into the core and config repos. do not deal with any specific plugins in the core code.

# 

please let us try to not intermix openproject's core and openproject-plugin-combinations (configs, in finnlabs speak) with each other any further.

for example: the core has its set of gems, therefore it has its very own `Gemfile.lock`. the core also has its clearly defined database schema, resulting from the set of migrations at the moment, therefore it has its very own `schema.rb`, which **should** be checked it.

configs contain plugins, which contain gem dependencies, therefore configs have their own `Gemfile.lock` which differs from the one of the core and should be checked into the `config`s repo. plugins also change the database by adding migrations, therefore the database schema differs from the core's. logically, there will be a different `schema.rb` which **should** be checked into the config's repo as well.

here, we have a rake task that deals with a file called `Gemfile.plugins`, which only exists in configs. it also references some specific plugins by name. therefore, the rake task as it is now should not live in the core (it could be callable from the project root by some means, though).

kgalli committed:
Remove logic to handle plugin dependencies

Plugin dependencies should be solved by bundler
only. Dependency logic was implemented due to
the fact that some OpenProject plugins do not
handle their dependencies properly.

Such issues have to be solved by
changing the source code of the plugins.

Therefore the white list of plugins (plugins.yml)
has been reduced to plugins which do not suffer
from dependency issues described above.

myabc commented:
@jonasheinrich this is targeting 4.2, right? (I set the milestone)

kgalli commented:
@myabc Yes would be great to see this in 4.2. I suggest we publish all plugins as gems on rubygems.org so this rake task can be implemented as easy as possible. At least in regard to resolving dependencies. I already removed the relatively complex dependency logic but it is still in progress.

jonasheinrich commented:
I'm gonna close this since it is not and won't be finished in the next time. Again, here is the relevant work package: https://community.openproject.org/work_packages/18720
